dags/iop/xml_extractors.py

```diff
@@ -0,0 +1,112 @@
+                article.find(self.source, self.prefixes).text
+            )
+            return node_with_prefix
+        node = article.find(self.source)
+        # if the value is required, it cannot be empty array or [None]
+        if bool(node) and all(node) and self.required:
+            return self.extra_function(article.find(self.source).text)
+        elif not self.required and node:
+            return self.extra_function(article.find(self.source).text)
+        elif self.required and node is None:
+            raise RequiredFieldNotFoundExtractionError(self.destination)
+        elif not bool(self.default_value) and self.required:
+            raise RequiredFieldNotFoundExtractionError(self.destination)
+        return self.default_value
+
+
+class AttributeExtractor(IExtractor):
+    def __init__(
+        self,
+        destination,
+        source,
+        attribute,
+        default_value=None,
+        required=False,
+        extra_function=lambda x: x,
+    ) -> None:
+        super().__init__(destination)
+        self.destination = destination
+        self.source = source
+        self.attribute = attribute
+        self.extra_function = extra_function
+        self.default_value = default_value
+        self.required = required
+
+    def extract(self, article: ET.Element):
+        value = self.extra_function(article.find(self.source).get(self.attribute))
+        # if the value is required, it cannot be empty array or [None]
+        if bool(value) and all(value) and self.required:
+            return value
+        elif not self.required and value:
+            return value
+        elif self.required and value is None:
+            raise RequiredFieldNotFoundExtractionError(self.destination)
+        elif not bool(self.default_value) and self.required:
+            raise RequiredFieldNotFoundExtractionError(self.destination)
+        return self.default_value
+
+
+class CustomExtractor(IExtractor):
+    def __init__(
+        self, destination, extraction_function, required=False, default_value=None
+    ) -> None:
+        super().__init__(destination)
+        self.destination = destination
+        self.extraction_function = extraction_function
+        self.default_value = default_value
+        self.required = required
+
+    def extract(self, article: ET.Element):
+        value = self.extraction_function(article)
+        # if the value is required, it cannot be empty array or [None]
+        if bool(value) and all(value) and self.required:
+            return value
+        elif not self.required and value:
+            return value
+        elif self.required and value is None:
+            raise RequiredFieldNotFoundExtractionError(self.destination)
+        elif not bool(self.default_value) and self.required:
+            raise RequiredFieldNotFoundExtractionError(self.destination)
+        return self.default_value
+
+
+class ConstantExtractor(IExtractor):
```

Comment:
where is this used?

Reply:
nowhere, it was since the begging when Ricardo wrote the extractor